Assets/PatchKit Patcher/Scripts/AppData/Remote/NodeTester.cs

```diff
@@ -1,147 +1,54 @@
-
-        private Thread _thread = null;
-
-        private DownloadSpeedCalculator _calculator = new DownloadSpeedCalculator();
-
-        private bool _isDone;
-
-        private bool _wasSuccess;
-
-        private static readonly ulong DefaultSize = 50;
-        private static readonly double DefaultSeed = 0.123;
-
-        private static readonly TimeSpan MaxTestDuration = TimeSpan.FromSeconds(15.0);
-
-        public static string PruneUrl(string url, ulong size, double seed)
-        {
-            var uri = new Uri(url);
-            return string.Format("{3}://{0}:8888/garbage.php?r={1}&ckSize={2}", uri.Host, seed, size, uri.Scheme);
-        }
-
-        public NodeTester(string url)
-            : this(url, DefaultSize, DefaultSeed)
-        {
+    public class NodeTester {
+        private static readonly TimeSpan MinTimeRemaining = TimeSpan.FromMinutes(2.0);
+        private static readonly TimeSpan MinDownloadTime = TimeSpan.FromSeconds(15.0);
+
+        private readonly ResourceUrl _resourceUrl;
+        private GarbageNodeTester _garbageNodeTester;
+        private Stopwatch _testingStopwatch;
+        private DownloadSpeedCalculator _downloadSpeedCalculator;
+
+        public NodeTester(ResourceUrl url) {
+            _resourceUrl = url;
+            _testingStopwatch = new Stopwatch();
+            _testingStopwatch.Start();
         }
 
-        public NodeTester(string url, ulong size, double seed)
-        {
-            if (string.IsNullOrEmpty(url))
-            {
-                throw new ArgumentException("url cannot be null or empty", "url");
-            }
+        public void Start(CancellationToken cancellationToken) {
+            _garbageNodeTester = new GarbageNodeTester(_resourceUrl.Url);
+            _garbageNodeTester.Start(cancellationToken);
 
-            _url = url;
-            _size = size;
-            _seed = seed;
+            _testingStopwatch.Stop();
         }
 
-        public bool IsReady
-        {
-            get
-            {
-                return _isDone;
-            }
-        }
-
-        public bool WasSuccess
-        {
-            get
-            {
-                return _wasSuccess;
-            }
+        public bool CanStart(long dataSize, DownloadSpeedCalculator calculator) {
+            return calculator.TimeRemaining(dataSize) > MinTimeRemaining
+                && _garbageNodeTester == null
+                && _testingStopwatch.IsRunning
+                && _testingStopwatch.Elapsed > MinDownloadTime;
+        } 
+
+        public bool IsReady {
+            get {
+                return _garbageNodeTester != null && _garbageNodeTester.IsReady;
+            } 
         }
 
-        public double BytesPerSecond
-        {
-            get
-            {
-                if (!_isDone || !_wasSuccess)
-                {
+        public double BytesPerSecond {
+            get {
+                if (_garbageNodeTester == null) {
                     return -1.0;
```

Comment:
1. Is it legal to read BytesPerSecond if IsDone returns false? It may be safer to throw illegal state exception here.
2. If you agree to 1, change the if to check `IsDone`.